[JPA - 3단계] 질문 삭제하기 리팩터링 리뷰 요청

oper912 opened the pull request:
안녕하세요 리뷰어님!

JPA 3단계 리뷰 요청드립니다.

이전에는 필요없을거같아 고려하지않았던 Question과 Answers의 양방향맵핑을 적용하였으며
QnAServiceTest 소스를 건드려도 되는건지는 확실하진않았으나 양방향 맵핑에 따라 불필요한 코드가 존재해서 삭제 및 수정하였습니다

잘 부탁드립니다 :smile:

oper912 committed:
refactor: CannotDeleteException RuntimeException 으로 변경

oper912 committed:
docs: application.properties 설정 값 정리

oper912 committed:
refactor: Question Answer 양방향 매핑 및 Answer Embedded 타입으로 선언

oper912 committed:
feat: 양방향 매핑용 Answer 일급 컬렉션 추가

oper912 committed:
test: Question 삭제 및 이력 반환 테스트

oper912 committed:
test: Answer 삭제 및 이력 반환 테스트

oper912 committed:
feat: Answer 삭제 로직 추가

oper912 committed:
feat: implement Answers

oper912 committed:
feat: DeleteHistory 필요한 getter 추가

oper912 committed:
test: 삭제 이력 일급컬렉션 삭제 가능 및 이력 반환 테스트

oper912 committed:
feat: 삭제 이력 일급컬렉션 추가

oper912 committed:
refactor: Question 소프트 삭제 로직 수정

oper912 committed:
feat: Question 답변들 삭제 로직 추가

oper912 committed:
test: QnAServiceTest 질문, 답변 양방향 매핑에 따른 불필요 코드 제거

oper912 committed:
refactor: QnAService 질문 삭제 로직 간소화

oper912 committed:
feat: Deleted wrapper class 추가

oper912 committed:
refactor: Deleted Embedded 타입 적용

oper912 committed:
feat: embedded 타입 클래스 기본 생성자 추가

kyucumber commented on src/main/java/qna/domain/Answer.java:
```diff
@@ -81,15 +98,15 @@ public String getContents() {
     }
 
     public void setDeleted(boolean deleted) {
-        this.deleted = deleted;
+        this.deleted.setDeleted(deleted);
     }
```
delete의 경우 항상 deleted의 값을 true로 할당하게 됩니다.

setDeleted가 아닌 delete 등의 메소드를 생성해주시는건 어떨까요?

oper912 replied:
값을 set해주는거보다는 리뷰어님 피드백대로 객체에 삭제를 하라는 메시지를 던지는게 맞는거같아 별도 메서드 추가하였습니다!

kyucumber commented on src/main/java/qna/domain/Answer.java:
```diff
@@ -64,6 +69,18 @@ public void toQuestion(Question question) {
         this.question = question;
     }
 
+    public DeleteHistory deleteAndReturnDeleteHistory(User loginUser) {
+        checkPossibleDelete(loginUser);
+        setDeleted(true);
+        return new DeleteHistory(ContentType.ANSWER, id, writer, LocalDateTime.now());
+    }
```
`DeleteHistory`의 참조가 `Answer` 내부에 들어가게 되었네요.

`Answer`에서 delete를 호출하면 `Answer` 내부에서 `DeleteHistory`를 반환하고 있습니다.

`DeleteHistory`의 경우 `Answer`, `Question` 등 늘어나는 도메인들 모두에서 사용될 가능성이 높아보입니다.

점점 객체들 간 결합이 많아질테고, 이후에는 `DeleteHistory` 하나의 변경으로 인해 다른 많은 객체들이 영향을 받을 가능성이 높아질 수 있지 않을까요?

이 부분이 문제라고 생각되지는 않지만 관련 부분에 대해 고민해주시면 좋을 것 같네요.

oper912 replied:
넵 알겠습니다!
그리고 기존에 QnAService에서 질문 데이터 삭제시 코드가 1번 코드와와 같았는데
코드를 다시 보니 질문 데이터를 삭제하면서 삭제 이력이 생기는게 아니라 삭제 이력을 생성하는 형식의 코드가 바람직하지 않은거 같아 2번으로 리팩토링하였습니다.

1번
```
    @Transactional
    public void deleteQuestion(User loginUser, Long questionId) {
        Question question = findQuestionById(questionId);
        DeleteHistories deleteHistories = DeleteHistories.of(loginUser, question);
        deleteHistoryService.saveAll(deleteHistories.getDeleteHistories());
    }
```

2번
```
    @Transactional
    public void deleteQuestion(User loginUser, Long questionId) {
        Question question = findQuestionById(questionId);
        deleteHistoryService.saveAll(question.deleteAndReturnDeleteHistories(loginUser));
    }
```

kyucumber commented on src/main/java/qna/domain/wrapper/Answers.java:
```diff
@@ -0,0 +1,34 @@
+    public Answers() {}
+
+    public void add(Answer answer) {
+        if (!answers.contains(answer)) {
+            this.answers.add(answer);
+        }
```
똑같은 답변이 또 들어오는것은 뭔가 잘못된 상황일 가능성이 높아 보입니다.

예외를 발생시키는건 어떨까요?

oper912 replied:
그러네요 똑같은 답변이 추가됐다는건 보통의 경우에는 발생하지 않아야 될 상황이기에 예외를 발생시키는게 맞는 것 같습니다

kyucumber commented on src/test/java/qna/domain/AnswerTest.java:
```diff
@@ -1,6 +1,48 @@
+import static org.assertj.core.api.Assertions.assertThatThrownBy;
+import static org.junit.jupiter.api.Assertions.assertAll;
+
 public class AnswerTest {
```
엔티티에 대한 테스트도 잘 작성해주셨네요 👍

kyucumber commented on src/test/java/qna/domain/AnswerTest.java:
```diff
@@ -1,6 +1,48 @@
+        DeleteHistory deleteHistory = answer.deleteAndReturnDeleteHistory(UserTest.JAVAJIGI);
+
+        // then
+        assertAll(
+                () -> assertThat(deleteHistory.getContentId()).isEqualTo(A1.getId()),
+                () -> assertThat(deleteHistory.getContentType()).isEqualTo(ContentType.ANSWER),
+                () -> assertThat(deleteHistory.getDeletedBy()).isEqualTo(A1.getWriter())
+        );
```
`LocalDateTime.now()`는 값 검증을 하지 않으셨네요.

이런 경우는 어떻게 테스트를 하면 좋을까요?

만약 어렵다면 값이 설정되어 null이 아닌지 정보 여부라도 체크해보시면 어떨까요?

oper912 replied:
이부분은 찾아보았는데 결국은 테스트하기 위해서는 날짜 데이터를 파라미터로 받을 수 있어야 하기때문에 현재 Question하고 Answer는 파라미터로 받지않고 내부적으로 LocalDateTime.now()를 사용하고 있어 테스트하기 어려워보입니다
지금은 null이 아닌지만 검증하도록 테스트 수정하도록 하겠습니다!

kyucumber reviewed (changes requested):
안녕하세요

몇가지 코멘트 남겨두었으니 확인 부탁드립니다.

oper912 committed:
refactor: 삭제시 deleted 상태 변경 delete 메서드를 통해 하도록 변경

소프트 삭제시 객체에게 삭제하라는 메시지를 전달하는 형태로 변경
https://github.com/next-step/jwp-qna/pull/91#discussion_r643239096

oper912 committed:
refactor: 질문에 동일한 답변 추가시 예외 발생하도록 수정

oper912 committed:
test: 답변 삭제시 삭제 이력에 시간 데이터 존재 여부 테스트 추가

oper912 committed:
refactor: 질문 삭제시 삭제 이력 일급컬렉션 반환 형태로 변경

kyucumber reviewed (approved):
피드백 잘 반영해주셨습니다.

다음 단계 진행을 위해 머지할게요
